crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -279,74 +231,97 @@ impl Plugin for VisibilityPlugin {
+            AssetEvent::Created { handle } | AssetEvent::Modified { handle } => {
+                if changed_aabbs.contains(handle) {
+                    // If the Aabb was manually added for this Handle<Mesh>, do not recompute it
+                    changed_aabbs.remove(handle);
+                    continue;
+                }
+                if let Some(mesh) = meshes.get(handle) {
+                    if let Some(updated_aabb) = mesh.compute_aabb() {
+                        mesh_aabb_map.insert(handle.clone_weak(), updated_aabb);
+                        updated_aabbs.insert(handle.clone_weak());
+                    }
+                }
+            }
+            AssetEvent::Removed { handle } => {
+                mesh_aabb_map.remove(handle);
             }
         }
     }
 
-    let to_update = |event: &AssetEvent<Mesh>| {
-        let handle = match event {
-            AssetEvent::Modified { handle } => handle,
-            _ => return None,
-        };
-        let mesh = meshes.get(handle)?;
-        let entities_with_handle = entity_mesh_map.entities_with_mesh.get(handle)?;
-        let aabb = mesh.compute_aabb()?;
-        Some((aabb, entities_with_handle))
-    };
-    for (aabb, entities_with_handle) in mesh_events.iter().filter_map(to_update) {
-        for entity in entities_with_handle {
-            commands.entity(*entity).insert(aabb.clone());
+    // Check and update Aabbs on entities with Handle<Mesh>. This handles the following cases:
+    // - The Handle<Mesh> on the entity is modified.
+    // - The Aabb corresponding to the Handle<Mesh> was updated. This will happen if a Mesh itself
+    //   is modified.
+    // - An Aabb was just added to or updated on the entity outside of this system.
+    for (entity, mesh_handle, mesh_changed, aabb) in queries.p1().iter_mut() {
+        // If an Aabb was just added/modified outside of this system, skip updating the component
+        if changed_aabbs.contains(mesh_handle) {
+            continue;
+        }
+        // If the Handle<Mesh> on the entity was modified, or the Aabb was updated due to the Mesh
+        // itself being modified, then update the entity's Aabb component
+        if mesh_changed || updated_aabbs.contains(mesh_handle) {
+            if let Some(updated_aabb) = mesh_aabb_map.get(mesh_handle) {
+                if let Some(mut aabb) = aabb {
+                    // If there is an existing Aabb component, we can update it in-place
+                    *aabb = updated_aabb.clone();
+                } else {
+                    commands.entity(entity).insert(updated_aabb.clone());
+                    // NOTE: Inserting a component via Commands causes a Changed<Component> query
+                    // in this system to be hit next time the system is run. As such, track the
+                    // Handle<Mesh> for which Aabbs were inserted on this run in order to avoid
+                    // detecting them as Changed on the next run.
```

Comment:
Would it be better to put this in a different system and stage? Using change detection on something modified in the same system should be avoided, in my experience.

Reply:
I'm not sure how that would help as the insertion would cause `Changed` to be triggered there too.

Reply:
If the system that inserts missing `Aabb` components runs at an earlier stage, `Changed<Components>` should be triggered on the same frame as the first case on line 315. Or, the `Aabb` would only be updated in place, and would definitely be changed on the frame that it is first calculated.

Although I prefer avoiding component insertion at arbitrary updates, and would prefer adding the `Aabb` in some bundle (`MaterialMeshBundle`?). This means there's no need to check if the component needs to be added, because if it is missing, that means the user didn't want that component for some reason.

Reply:
If it were part of the bundle, we’d have to add something else to Aabb to indicate that it is up to date or something. If Aabbs are added with a default bound then we would have to compute it to update it to be correct. How would we know that the Aabb is incorrect without calculating it and comparing?

Reply:
That brings up something else: an `Aabb` as currently defined always has a center, but the bounds a `Mesh` with no vertices has no center and zero half extents, while a `Mesh` with a single vertex does have a center (the location of the vertex) and zero half extents.

I use:
```rust
pub enum Bounds<A> {
    Zero,
    Bounds { lo: A, hi: A },
}
```
so that I can use `Bounds::Zero` as the default (the bounds of nothing), which allows me to properly join any two bounds, because joining `Zero` to another bounds should be a no-op. (`Zero` with joining forms a monoid, if you're familiar with the concept.)

Then I have:
```rust
pub struct DefinedBounds<A> {
    pub lo: A,
    pub hi: A,
}
```
when I know there is at least one point inside the bounds.

I suggest modifying `Aabb` to also have a zero point case that can serve as the default. Then you can detect that it needs updating if the `Mesh` has positions, and the default is correct until the mesh is loaded.

Reply:
What is the purpose of a `Mesh` with no vertices? Also, it sounds like you're addressing the 'uninitialised' case, but what I was suggesting was the case of a `Mesh` that has been updated and maybe or maybe not the `Aabb` was updated this frame too. That is, the `Aabb` has been initialised but is perhaps no longer correct. I suppose your earlier argument is - should manual `Aabb` definition be allowed outside the bounds of `NoAabbUpdate`. If not, that would simplify things for sure.

Reply:
Can we move this to use an event system to make updates unambiguous?

- Newtype the Aabb over a private AabbInner type
- Add a public getter that returns the inner AabbInner data
- Add an event `UpdateAabb`, which is now the only public way to update an `Aabb`
- When a mesh has updated, you can check if an event was sent to update the Aabb. If not, go ahead and compute it and send your own event.

Reply:
That or a generational marker. Every time a mesh is updated, bump the generation on the mesh. Check if the Aabb has the same generation as the mesh. If not, update it. External updates would need to bump the generation to match the mesh.

Reply:
> What is the purpose of a Mesh with no vertices?

A `Mesh` with no vertices may not be used intentionally, but generated procedurally. For example, subtracting a shape from another that the first shape encloses results in an empty shape. I could remove the `Mesh` component or something, but the most natural representation is an empty `Mesh`. Also, adding and removing components can play havoc with other systems.

Furthermore, an empty mesh is a reasonable thing that has reasonable behavior, so handling it properly makes Bevy more robust, and avoids special cases that must be handled by the user. It's also usually worth implementing no-op behavior e.g. if you want to join several `Aabb`s, what do you start with? When summing integers, you can start with 0 (the identity element for addition). This extends to, say, filtering based on visibility. `Option<Aabb>` would just be hacking on an identity element without clear semantics - is `None` the identity, or an error, or maybe an infinite `Aabb`?

I do think that there needs to be a clear indication of which system should be updating the `Aabb`, because otherwise it could be a race.